demos/python_demos/asl_recognition_demo/asl_recognition_demo/tracker.py

```diff
@@ -0,0 +1,222 @@
+"""
+ Copyright (c) 2019 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+from scipy.optimize import linear_sum_assignment
+
+
+class Detection:
+    """Class that stores detected object"""
+
+    def __init__(self, obj_id, roi, conf, waiting=0, duration=1):
+        """Constructor"""
+
+        self.id = obj_id
+        self.roi = roi
+        self.conf = conf
+        self.waiting = waiting
+        self.duration = duration
+
+    @property
+    def roi(self):
+        """Returns ROI of detected object"""
+
+        return self._roi
+
+    @roi.setter
+    def roi(self, roi):
+        """Sets ROI of detected object"""
+
+        self._roi = np.copy(roi.reshape(1, -1))
+
+
+class Tracker:  # pylint: disable=too-few-public-methods
+    """Class that carries out tracking of persons using Hungarian algorithm"""
+
+    def __init__(self, detector, score_threshold, iou_threshold, smooth_weight=0.5, max_waiting=5):
+        """Constructor"""
+
+        self._detector = detector
+        self._score_threshold = score_threshold
+        self._iou_threshold = iou_threshold
+        self._smooth_weight = smooth_weight
+        self._max_waiting = max_waiting
+
+        self._last_detections = []
+        self._cur_req_id, self._next_req_id = 0, 1
+        self._last_id = 0
+
+    @staticmethod
+    def _matrix_iou(set_a, set_b):
+        """Computes IoU metric for the two sets of vectors"""
+
+        intersect_ymin = np.maximum(set_a[:, 0].reshape([-1, 1]), set_b[:, 0].reshape([1, -1]))
+        intersect_xmin = np.maximum(set_a[:, 1].reshape([-1, 1]), set_b[:, 1].reshape([1, -1]))
+        intersect_ymax = np.minimum(set_a[:, 2].reshape([-1, 1]), set_b[:, 2].reshape([1, -1]))
+        intersect_xmax = np.minimum(set_a[:, 3].reshape([-1, 1]), set_b[:, 3].reshape([1, -1]))
+
+        intersect_heights = np.maximum(0.0, intersect_ymax - intersect_ymin)
+        intersect_widths = np.maximum(0.0, intersect_xmax - intersect_xmin)
+
+        intersect_areas = intersect_heights * intersect_widths
+        areas_set_a = ((set_a[:, 2] - set_a[:, 0]) * (set_a[:, 3] - set_a[:, 1])).reshape([-1, 1])
+        areas_set_b = ((set_b[:, 2] - set_b[:, 0]) * (set_b[:, 3] - set_b[:, 1])).reshape([1, -1])
+
+        areas_set_a[np.less(areas_set_a, 0.0)] = 0.0
+        areas_set_b[np.less(areas_set_b, 0.0)] = 0.0
```

Comment:
1. It can be replaced with clip: `areas_set_a.clip(0.0, out=areas_set_a)`

2. These lines make me think that it is possible that `set_a[:, 2] - set_a[:, 0] < 0` or `set_a[:, 3] - set_a[:, 1] < 0` but that means that `set_a[:, 2]` and `set_a[:, 0]` should be just swapped (or `abs()` should be used) to get correct results instead of pretending that the rect is a line or a dot.

Reply:
Regarding the second item, the negative values of `set_a[:, 2] - set_a[:, 0]` or `set_a[:, 3] - set_a[:, 1]` are Ok and mean that bboxes are not intersected and therefore metric value is `0`.

Reply:
Are bboxes inside of one set intersected with each other? I think your aim is to intersect `set_a` with `set_b`. `set_a[:, 2] - set_a[:, 0]` seems to be computing width (or height).

Reply:
Oh, yes, I've mixed up it. But I prefer to save this logic due to possible mistakes in the bbox prediction part (I'm not sure that it can appear with current detector but some other detectors can filter boxes by setting negative limits).

Reply:
This is different: you should clip `set_a` than, but not width. And control not only the lower bound but the upper bound as well.

Reply:
No, negative values are the intentional marker of incorrect detections (in case when we don't want to change the confidence score for any reason).

Reply:
> I prefer to save this logic due to possible mistakes in the bbox prediction part
negative values are the intentional marker of incorrect detections

So are negative values mistakes or intentional? Or that phrases refer to different objects?

If negative values can be intentional predictions of some detectors, will you point me out where I can learn about that idea?

Reply:
An author of a network can do everything (including negative bbox values if it can speed up the future post-processing stage) but in this situation I agree it's overhead. Fixed on `abs`.

Reply:
The similar update should be done for `areas_set_b`.

But I have realized this `abs` doesn't solve the problem completely because another part of function assumes that `set_a[:, 0]` is upper edge of bbox, `set_a[:, 1]` is left edge of bbox and so on. This comes from `intersect_ymin = np.maximum(set_a[:, 0].reshape([-1, 1]), set_b[:, 0].reshape([1, -1]))`. If `set_a[:, 0]` was lower edge, it would be `intersect_ymin = np.MINimum(set_a[:, 0].reshape([-1, 1]), set_b[:, 0].reshape([1, -1]))`.

I think it can be easier to swap elements if they are in wrong order (or reject such objects) in the very beginning and then remove `intersect_heights = np.maximum(0.0, intersect_ymax - intersect_ymin)`, `intersect_widths = np.maximum(0.0, intersect_xmax - intersect_xmin)`, `iou_values[np.less_equal(union_areas, 0.0)] = 0.0` lines and `abs`.

Actually current implementation already sets IoU for bboxes with wrong vertex order to 0 because of that `intersect_heights = np.maximum(0.0, intersect_ymax - intersect_ymin)`. So `abs` is not needed in that approach too as well as `iou_values[np.less_equal(union_areas, 0.0)] = 0.0` because IoU will be 0 any way. But that implementations looks tricky and unclear.

Reply:
I've added bbox filtering before the call of IoU function, so there is no need in processing invalid sizes. Regarding the suggestion to remove `intersect_heights = np.maximum(0.0, intersect_ymax - intersect_ymin), intersect_widths = np.maximum(0.0, intersect_xmax - intersect_xmin)` lines (if I correctly understand the suggestion) - negative values mean that there is not intersection... , so we replace it on zeros values to make the final metric for this boxes to be zero too.

Reply:
You are right